### ISSUE TITLE
Input method, read-only buffers and Avy

### ISSUE THREAD

**Issue by zanzellor** [abo-abo/avy#197]
I've got english-dvorak set as my input method (M-x set-input-method RET english-dvorak). In a normal buffer, when I run avy-goto-char (or any other jump command), it lets me type the char to jump to using the dvorak layout, and uses the QWERTY layout for the decision chars, which is the desired behaviour.
The problem occurs when a buffer is read-only. When i run a jump command in a read-only buffer, instead of using the set input method (english-dvorak in my case), it uses QWERTY for the char/word to jump to.
Toggling between the buffer's read-only state changes the behavior back and forth between the ones described above.
To reproduce, in a normal buffer, set any input method besides the default (such as 'english-dvorak') and try to jump to any char. You'll see that to type the char to jump to, avy uses the newly set input method's layout. Now set the buffer as read-only (M-x read-only-mode), and try jumping to a char again. This time, avy uses the default input method's layout instead of the one set by the user.

**Comment by abo-abo** [comment]
I can reproduce the issue, but it's not avy's fault. Avy uses:

    (read-char "char: " t)

Where `t` represents `inherit-input-method`. When the buffer is not read-only, it respects the input method. But when it is read-only, it doesn't.  I suggest you report this issue against base Emacs as a `read-char` bug.

> I've got english-dvorak set as my input method (M-x set-input-method RET english-dvorak)

As a work-around, since you call `set-input-method` yourself, you can add code to turn off `read-only-mode`. Or you can add an advice around `avy-goto-char` to disable `read-only-mode` for its duration.

**Comment by mathiasdahl** [comment]
Hi, I have what I think is a similar (if not the same) issue, so I hope it is okay if I hijack this issue a bit... :)

I am trying out avy-goto-word-1 right now and I like it. But, I am sometimes using the input method swedish-postfix (so that I can type Swedish characters like å, ä and ö with an English keyboard layout in my OS) and they don't play along well. If I use avy-goto-word-1 it asks for a char, and if I type any of a, o or e, Emacs will wait for more input in order to know what I want to type. Whatever I type then I get an error.

Above you mention this is a bug with Emacs, but when I try to evaluate, say (read-char "Char: "), it correctly returns back 111 when I type "o", so it seems read-char is working fine.

Any idea how I can fix this?

To recreate you should only need to enable the input method swedish-postfix and try to search for a word starting with "o".

Thanks!

/Mathias

**Comment by abo-abo** [comment]
> when I try to evaluate, say (read-char "Char: ")

Try evaluating `(read-char "Char: " t)`, since that's what avy does. When you do `(read-char "Char: ")`, the input method is not inherited, and regular ASCII is used.

I get something strange when I eval `read-char` like that and press `oo`: I get and underscored `o` inserted into the buffer. That should not happen at all.

**Comment by mathiasdahl** [comment]
Thanks! Seems like read-char and input methods (at least the one I use, and perhaps it applies to all -postfix input methods) does not play along very well. I guess one problem that can never really be solved is when I just want to type an `o`. In normal text typing, there will eventually be another key typed afterwards, but what should read-char do in that case? If I type `o` it still waits for input, but if I don't want to give more input, since I just want an `o`, what do I do, and how should read-char know what to do? Feels tricky, but I haven't thought about this for very long... :)

Am I right to assume that, yes, you want to use that optional parameter to read-char (to support other input methods that might work better)?

I of course have the option not to use avy's goto commands when I have an input method set, or the other way around.

**Comment by abo-abo** [comment]
> Am I right to assume that, yes, you want to use that optional parameter to read-char (to support other input methods that might work better)?

Yes, it works fine if I set the input method to `ukrainian-computer`: the layout simply changes to Cyrillic. There are no letters that require double input, so it works perfectly.

**Comment by abo-abo** [comment]
You should `report-emacs-bug` for the combination of `swedish-postfix` and `read-char`.

**Comment by mathiasdahl** [comment]
Perhaps I will do that, or perhaps I will just "patch" my local version of avy :) since I can live without being able to goto Swedish characters when I have an English keyboard layout and use swedish-postfix.

Thanks for your prompt comments!
